Added retry logic if client request returns an error code that is configured for retrying

criminosis opened the pull request:
Closes #4408

To wait between retries since the code being called into Elasticsearch's client appears to be the synchronous API I opted for just using `Thread.sleep`. If there's a better utility or if a different approach is desired happy change.

Additionally I marked the new configuration options with the `LOCAL` flag. I noticed `retry_on_conflict` went with `MASKABLE` but it didn't seem readily apparent what further obligations would be needed to fulfill the "global" portion of that contract, so just figured it'd be best to start with `LOCAL` and ask about it in the PR.

---
Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there an issue associated with this PR? Is it referenced in the commit message?
- [X] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [X] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [X] Is your initial contribution a single, squashed commit?

### For code changes:
- [X] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [X] Have you ensured that format looks appropriate for the output in which it is rendered?

li-boxuan reviewed (commented):
Thanks for taking care of this! My two cents:

1) Either local or maskable is fine. Maskable means the change will be (automatically) propagated to different instances, but every instance could override the value locally. Local is simpler and has less overhead (as no propagation is needed). I would prefer to use LOCAL here.
2) It's okay to write our own exponential backoff code.

criminosis commented on janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java:
```diff
@@ -546,13 +572,42 @@ private Response performRequest(String method, String path, byte[] requestData)
+        } while (retryCount <= retryAttemptLimit);
+
+        //If we've exceeded our retry limit then throw the last exception that was captured from retrying
+        throw lastException;
```
I don't follow how it's a bug @li-boxuan.

> If the delegate.performRequest(request) does not throw a ResponseException on the first attempt and there are no retry attempts (because the response is successful or an error not suitable for retry occurs), the do-while loop will not enter or will exit early, and lastException will never be assigned a value.

ChatGPT seems to have missed the semantic implication that other exception types are not caught and thus would be thrown _immediately_ out of the method. In that scenario `lastException` never being assigned a value is moot.

The four cases should be:

1. The request is successful - we return out of the method.
2. The request is not successful - but it's an exception that's different than a `ResponseException` so it immediately gets thrown up out of the method.
3. The request is not successful - it is a `ResponseException` _but_ it is not a configured error code for retry and thus will be immediately thrown up out of the method.
4. The request is not successful - it is a `ResponseException` _and_ it *is* a configured error code for retry, which will then cause `lastException` to be assigned and possibly later thrown if it is the last (or only) attempt if retries are set to 0.

criminosis replied:
Added some more tests in 9b57c0349 to verify what ChatGPT is claiming is a non-issue.

li-boxuan replied:
Oh lol thank you so much for the detailed explanation. I rely on AI too much these days... and got cheated by its hallucination that at the first glance, seemed reasonable.

criminosis commented on janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java:
```diff
@@ -304,6 +304,26 @@ public class ElasticSearchIndex implements IndexProvider {
+
+    public static final ConfigOption<String[]> RETRY_ERROR_CODES =
+        new ConfigOption<>(ELASTICSEARCH_NS, "retry-error-codes",
+            "Elasticsearch REST client ResponseException error codes To Retry.", ConfigOption.Type.LOCAL,
```
Added in 5354c7dc7 👍

li-boxuan reviewed (approved):
This is great, thanks!

porunov commented on janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java:
```diff
@@ -419,6 +429,22 @@ public void setRetryOnConflict(Integer retryOnConflict) {
+        this.retryMaxWaitMs = retryMaxWaitMs;
+    }
+
+    public void setRetryOnErrorCodes(Set<Integer> retryOnErrorCodes) {
+        this.retryOnErrorCodes.addAll(retryOnErrorCodes);
+    }
```
The method name is `setRetryOnErrorCodes`, but instead of `set` operation it `adds`.
You can fix it by actually setting `this.retryOnErrorCodes = retryOnErrorCodes` instead of mutating existing collection.
However, it seems there is no need to execute this method more than once during the lifetime of this class.
Thus, I would suggest moving all of these parameters into the constructor instead. I know that `setRetryOnConflict` and `setBulkRefresh` have the same problem, but I think it should be refactored.

Feel free to ignore this comment if you see any complications in this refactoring.

criminosis replied:
Happy to make the change but for what it's worth that's going to turn the signature into:

```
public RestElasticSearchClient(RestClient delegate, int scrollKeepAlive, boolean useMappingTypesForES7,
                                   int retryAttemptLimit, Set<Integer> retryOnErrorCodes, long retryInitialWaitMs,
                                   long retryMaxWaitMs) {
```

Which may be pushing a parameter parameter count for transitioning to a builder. Especially if `setRetryOnConflict` and `setBulkRefresh` are added one day.

criminosis replied:
@porunov Moved the new configuration settings into the constructor as suggested. Isolated that into commit c3961fc22. 

If that looks acceptable then I'll do the squash & sign-off to get the PR into a mergable state again. Figured I'd wait before squashing it in case you had second thoughts upon seeing it given my aforementioned comment.

porunov replied:
> Happy to make the change but for what it's worth that's going to turn the signature into:
> 
> ```
> public RestElasticSearchClient(RestClient delegate, int scrollKeepAlive, boolean useMappingTypesForES7,
>                                    int retryAttemptLimit, Set<Integer> retryOnErrorCodes, long retryInitialWaitMs,
>                                    long retryMaxWaitMs) {
> ```
> 
> Which may be pushing a parameter parameter count for transitioning to a builder. Especially if `setRetryOnConflict` and `setBulkRefresh` are added one day.

Yes, that's a refactoring issue, I agree. However, I think it still better to have a clear contract for this class creation. Later on this contract can be simplified by moving all these parameters to a separate "configuration" POJO for easier extensibility and clearer constructor.
I like it better how it looks right now than it was before. Thank you!

porunov commented on janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java:
```diff
@@ -546,13 +572,42 @@ private Response performRequest(String method, String path, byte[] requestData)
+        int retryCount = 0;
+        ResponseException lastException;
+        do {
+            if (retryCount > 0) {
+                //Wait before trying again
+                long waitDurationMs = Math.min((long) (retryInitialWaitMs * Math.pow(10, retryCount - 1)), retryMaxWaitMs);
+                log.warn("Retrying Elasticsearch request in {} ms. Attempt {} of {}", waitDurationMs, retryCount, retryAttemptLimit);
+                try {
+                    Thread.sleep(waitDurationMs);
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(String.format("Thread interrupted while waiting for retry attempt %d of %d", retryCount, retryAttemptLimit), e);
+                }
+            }
+            try {
+                return delegate.performRequest(request);
+            } catch (ResponseException e) {
+                if (!retryOnErrorCodes.contains(e.getResponse().getStatusLine().getStatusCode())) {
+                    throw e;
+                }
+                lastException = e;
+            }
```
it seems logic under the if statement `if (retryCount > 0)` needs to be executed on exception only. I would suggest moving the whole logic under that if statement under the `catch`. In this case you avoid having `if (retryCount > 0)`.
I.e. : 
```
} catch (ResponseException e) {
  if (!retryOnErrorCodes.contains(e.getResponse().getStatusLine().getStatusCode())) {
    throw e;
  }
  lastException = e;
  // put here that `sleep` logic.
}
```

criminosis replied:
I was going to comment back that doing so would mean we'd do a wait when we'd end up breaking out of the loop anyways if we don't have that check. But I realized we could condense the loop's condition into the `if` statement and then get the best of both worlds at the trade off of doing the loop as a `while (true)`.

Done in 13e65bf87

porunov replied:
Thank you. Looks great!

porunov reviewed (commented):
Thank you @criminosis ! I have just some nitpicks below. Otherwise looks good.

porunov reviewed (commented):
@criminosis please, feel free to squash the commits. However, notice that ElasticSearch tests are now failing:
```
Failures: 
Error:    RestClientSetupTest.testConnectBasicHttpConfigurationAllOptions:257 expected: <[408, 429]> but was: <null>
Error:  Errors: 
Error:    RestClientSetupTest.testSSLDisableHostNameVerifierDefaultOff » InvalidUseOfMatchers 
```

criminosis commented:
@porunov yeah, I just focused on compilation passing to present the new signatures in case you had feedback or further modification, etc.

Fixed it all up, squashed and re-signed off again. Should be ready for review again 👍

porunov reviewed (approved):
LGTM. Thank you for your contribution @criminosis !

criminosis commented:
Thank you @porunov & @li-boxuan , glad to finally get to find something to contribute back to JanusGraph!

Out of curiosity I see it's been earmarked for the 1.1.0 milestone. Assuming the build completely passes and it gets merged will it go out into nightly build / release prior to the 1.1.0 release?

Looking forward to getting this my cluster to see if it helps the issue that originally prompted this.

porunov commented:
> Thank you @porunov & @li-boxuan , glad to finally get to find something to contribute back to JanusGraph!
> 
> Out of curiosity I see it's been earmarked for the 1.1.0 milestone. Assuming the build completely passes and it gets merged will it go out into nightly build / release prior to the 1.1.0 release?
> 
> Looking forward to getting this my cluster to see if it helps the issue that originally prompted this.

Yes. If everything passes the snapshot release is automatically created for each merged commit for any supported branches (`master` and `v1.0` at the moment). You can read about snapshot releases here: https://docs.janusgraph.org/advanced-topics/commit-releases/
